### PR TITLE
[SYCL] Add pass that remove compiler internal metadata

### PR DESCRIPTION
**Pull request by MrSidims**
Right now the pass removes only sycl_aspects and
sycl_types_that_use_aspects module-level metadata, so during llvm-link these metadata nodes won't be duplicated significantly increasing module size. 

Other metadata like sycl_used_aspects (applied to a function) were also considered to be added to the pass, but at this moment such metadata causes no harm, so it was left untouched.

### CODE REVIEW

File: sycl/test/extensions/properties/properties_kernel_device_has.cpp [hunk 1]
```diff
@@ -37,90 +36,65 @@ int main() {
   auto Redu1 = reduction<int>(nullptr, plus<int>());
   auto Redu2 = reduction<float>(nullptr, multiplies<float>());
 
-  // CHECK-IR: spir_kernel void @{{.*}}WGSizeKernel0(){{.*}} #[[DHAttr1:[0-9]+]]
```
Comment (MrSidims) [inline sycl/test/extensions/properties/properties_kernel_device_has.cpp:40]: These checks (as well as the tests below) were removed as no longer relevant. This functionality is tested in `llvm/tests/SYCLLowerIR/PropagateAspectsUsage/*` , yet I'm as bit concerned by the change I make. If you have suggestions on how to keep the tests - I'm open for proposals.

File: clang/lib/CodeGen/BackendUtil.cpp [hunk 2]
```diff
@@ -1100,6 +1101,8 @@ void EmitAssemblyHelper::RunOptimizationPipeline(
       // Rerun aspect propagation without warning diagnostics.
       MPM.addPass(SYCLPropagateAspectsUsagePass(/*ExcludeAspects=*/{},
                                                 /*ValidateAspects=*/false));
+      // Remove SYCL metadata added by sycl-post-link tool, like sycl_aspects
+      MPM.addPass(CleanupSYCLCompilerInternalMetadataPass());
```
Comment (steffenlarsen) [inline clang/lib/CodeGen/BackendUtil.cpp:1105]: Could we do it after `CompileTimePropertiesPass` in case we want to clean something after that too? I could see that happening potentially.

File: llvm/include/llvm/SYCLLowerIR/CleanupSYCLCompilerInternalMetadata.h [hunk 3]
```diff
@@ -0,0 +1,31 @@
+//===- CleanupSYCLCompilerInternalMetadata.h - SYCLRemoveAspectsMD Pass --===//
```
Comment (asudarsa) [inline llvm/include/llvm/SYCLLowerIR/CleanupSYCLCompilerInternalMetadata.h:1]:
```suggestion
//===- CleanupSYCLCompilerInternalMetadata.h - CleanupSYCLCompilerInternalMetadata Pass --===//
```

File: clang/lib/CodeGen/BackendUtil.cpp [hunk 4]
```diff
@@ -43,6 +43,7 @@
 #include "llvm/Passes/PassBuilder.h"
 #include "llvm/Passes/PassPlugin.h"
 #include "llvm/Passes/StandardInstrumentations.h"
+#include "llvm/SYCLLowerIR/CleanupSYCLCompilerInternalMetadata.h"
```
Comment (asudarsa) [inline clang/lib/CodeGen/BackendUtil.cpp:46]: Nit. For sake of simplicity, this can be CleanupSYCLMetadata?

File: clang/lib/CodeGen/BackendUtil.cpp [hunk 5]
```diff
@@ -1100,6 +1101,8 @@ void EmitAssemblyHelper::RunOptimizationPipeline(
       // Rerun aspect propagation without warning diagnostics.
       MPM.addPass(SYCLPropagateAspectsUsagePass(/*ExcludeAspects=*/{},
                                                 /*ValidateAspects=*/false));
+      // Remove SYCL metadata added by sycl-post-link tool, like sycl_aspects
```
Comment (asudarsa) [inline clang/lib/CodeGen/BackendUtil.cpp:1104]: Metadata is added by front-end, if I am not mistaken.

File: llvm/lib/Passes/PassBuilder.cpp [hunk 6]
```diff
@@ -243,8 +244,8 @@
 #include "llvm/Transforms/Utils/CanonicalizeAliases.h"
 #include "llvm/Transforms/Utils/CanonicalizeFreezeInLoops.h"
 #include "llvm/Transforms/Utils/CountVisits.h"
-#include "llvm/Transforms/Utils/Debugify.h"
 #include "llvm/Transforms/Utils/DXILUpgrade.h"
+#include "llvm/Transforms/Utils/Debugify.h"
```
Comment (asudarsa) [inline llvm/lib/Passes/PassBuilder.cpp:248]: Why is this change needed? Is this something to keep format checkers happy?

**Comment by maksimsab** [comment]
`optional_kernel_features/*` tests look useful. Probably, we should introduce additional command line option to keep metadata for testing purposes.

**Review by asudarsa — APPROVED** [review]
Main changes look good to me. I will let test authors to comment on whether the test removals are valid. Few changes in code comments requested.

Thanks

File: llvm/lib/Passes/PassBuilder.cpp (hunk 6 above)
Comment (MrSidims) [reply llvm/lib/Passes/PassBuilder.cpp:248]: Just because of that, I copied its suggestion without checking. I'll revert it back to stay aligned with llvm.org

File: clang/lib/CodeGen/BackendUtil.cpp (hunk 5 above)
Comment (MrSidims) [reply clang/lib/CodeGen/BackendUtil.cpp:1104]: Yeah, my mistake, other metadata like sycl_used/declared_aspects are introduced by sycl-post-link

File: llvm/include/llvm/SYCLLowerIR/CleanupSYCLCompilerInternalMetadata.h [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+//
+//===----------------------------------------------------------------------===//
+//
+// Cleanup SYCL compiler internal metadata (usually inserted by sycl-post-link)
+// as it will never be used in the compilation ever again
+//
+//===----------------------------------------------------------------------===//
+//
+#ifndef LLVM_CLEANUP_SYCL_COMPILER_INTERNAL_METADATA
+#define LLVM_CLEANUP_SYCL_COMPILER_INTERNAL_METADATA
```
Comment (sarnex) [inline llvm/include/llvm/SYCLLowerIR/CleanupSYCLCompilerInternalMetadata.h:15]: Can we use `#pragma once`, some of the other headers in this folder are using it.

File: llvm/include/llvm/SYCLLowerIR/CleanupSYCLCompilerInternalMetadata.h [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+
+#include "llvm/IR/PassManager.h"
+
+#include <set>
```
Comment (sarnex) [inline llvm/include/llvm/SYCLLowerIR/CleanupSYCLCompilerInternalMetadata.h:19]: Do we need this include in the header?

File: llvm/lib/SYCLLowerIR/CleanupSYCLCompilerInternalMetadata.cpp [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+                                             ModuleAnalysisManager &MAM) {
+  // Remove SYCL module-level metadata that will never be used again to avoid
+  // its duplication of their operands during llvm-link hence preventing
+  // increase of the module size
+  llvm::SmallVector<llvm::StringRef, 2> ModuleMDToRemove = {
+      "sycl_aspects", "sycl_types_that_use_aspects"};
+  for (const auto &MD : ModuleMDToRemove)
+    cleanupSYCLCompilerModuleMetadata(M, MD);
+
+  return PreservedAnalyses::all();
```
Comment (sarnex) [inline llvm/lib/SYCLLowerIR/CleanupSYCLCompilerInternalMetadata.cpp:45]: I'm not an expert here but usually I see we keep track of if we actually did anything, and if so, we return `PreservedAnalyses::none()`, and if we did nothing, we return `PreservedAnalyses::all()`. In this case we only change metadata so I'm not sure if that is guaranteed to not modify any analyses.

File: llvm/lib/SYCLLowerIR/CMakeLists.txt [hunk 10]
```diff
@@ -64,6 +64,7 @@ add_llvm_component_library(LLVMSYCLLowerIR
   MutatePrintfAddrspace.cpp
   SYCLAddOptLevelAttribute.cpp
   SYCLPropagateAspectsUsage.cpp
+  CleanupSYCLCompilerInternalMetadata.cpp
```
Comment (sarnex) [inline llvm/lib/SYCLLowerIR/CMakeLists.txt:67]: Nit: should this be alphabetically sorted?

File: clang/lib/CodeGen/BackendUtil.cpp (hunk 4 above)
Comment (sarnex) [inline clang/lib/CodeGen/BackendUtil.cpp:46]: Do you mind giving some more information on the problem we are trying to solve?

File: llvm/lib/SYCLLowerIR/CleanupSYCLCompilerInternalMetadata.cpp [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+
+} // anonymous namespace
+
+PreservedAnalyses
+CleanupSYCLCompilerInternalMetadataPass::run(Module &M,
+                                             ModuleAnalysisManager &MAM) {
+  // Remove SYCL module-level metadata that will never be used again to avoid
+  // its duplication of their operands during llvm-link hence preventing
+  // increase of the module size
+  llvm::SmallVector<llvm::StringRef, 2> ModuleMDToRemove = {
```
Comment (sarnex) [inline llvm/lib/SYCLLowerIR/CleanupSYCLCompilerInternalMetadata.cpp:40]: In this case it seems both of these are used by SYCLPropagateAspectsUsage.cpp so maybe we should just delete the metadata there, but in case we need to add stuff in the future I think a separate pass makes more sense.

File: sycl/test/extensions/properties/properties_kernel_device_has.cpp (hunk 1 above)
Comment (sarnex) [reply sycl/test/extensions/properties/properties_kernel_device_has.cpp:40]: Yeah I'm a bit worried with the test removals. My only idea is to add a hidden -mllvm option or something to run the pass or not.

File: clang/lib/CodeGen/BackendUtil.cpp (hunk 4 above)
Comment (MrSidims) [reply clang/lib/CodeGen/BackendUtil.cpp:46]: May be, but there are probably other metadata that we would like not to clear

Comment (MrSidims) [reply clang/lib/CodeGen/BackendUtil.cpp:46]: The problem is the following:
We split IR module multiple times in the code split and then trying to link the split modules using llvm-link. Since llvm-link tool just regularizes the metadata it creates a multiple copies of the same module metadata increasing module size considerably.

File: llvm/include/llvm/SYCLLowerIR/CleanupSYCLCompilerInternalMetadata.h (hunk 8 above)
Comment (MrSidims) [reply llvm/include/llvm/SYCLLowerIR/CleanupSYCLCompilerInternalMetadata.h:19]: nope, thanks!

File: llvm/include/llvm/SYCLLowerIR/CleanupSYCLCompilerInternalMetadata.h (hunk 7 above)
Comment (MrSidims) [reply llvm/include/llvm/SYCLLowerIR/CleanupSYCLCompilerInternalMetadata.h:15]: we indeed can, but from what I can tell guarding macro is more standard practice in LLVM. For sure I won't fight for this and can change if you believe it's better :)

File: llvm/lib/SYCLLowerIR/CleanupSYCLCompilerInternalMetadata.cpp (hunk 11 above)
Comment (MrSidims) [reply llvm/lib/SYCLLowerIR/CleanupSYCLCompilerInternalMetadata.cpp:40]: There is another reason: SYCLPropagateAspectsUsage is being called 2 times. There is a way to ask it how many times it was called and do cleanup on the second call, but it seem to be fragile

File: llvm/lib/SYCLLowerIR/CleanupSYCLCompilerInternalMetadata.cpp (hunk 9 above)
Comment (MrSidims) [reply llvm/lib/SYCLLowerIR/CleanupSYCLCompilerInternalMetadata.cpp:45]: Thanks, I'm neither an expert, let me double check

File: sycl/test/extensions/properties/properties_kernel_device_has.cpp (hunk 1 above)
Comment (MrSidims) [reply sycl/test/extensions/properties/properties_kernel_device_has.cpp:40]: Well, unfortunately it won't work as opt doesn't provide a way to exclude one pass from the pipeline.

File: clang/lib/CodeGen/BackendUtil.cpp (hunk 4 above)
Comment (elizabethandrews) [reply clang/lib/CodeGen/BackendUtil.cpp:46]: I think that is ok. We aren't clearing all internal metadata as well right?

**Review by elizabethandrews — APPROVED** [review]
Other than nit about pass name, FE changes LGTM

File: clang/lib/CodeGen/BackendUtil.cpp [hunk 12]
```diff
@@ -1121,6 +1122,9 @@ void EmitAssemblyHelper::RunOptimizationPipeline(
       // Process properties and annotations
       MPM.addPass(CompileTimePropertiesPass());
 
+      // Remove SYCL metadata added by the frontend, like sycl_aspects
```
Comment (elizabethandrews) [inline clang/lib/CodeGen/BackendUtil.cpp:1125]: Nit: `Remove sycl_aspects and sycl_types_that_use_aspects module-level metadata`

I think being specific is helpful in the comments. It might also make sense to elaborate on why we are removing them

**Comment by AlexeySachkov** [comment]
> sycl_types_that_use_aspects module-level metadata, so during llvm-link these metadata nodes won't be duplicated significantly increasing module size. **Meanwhile these metadata aren't needed after sycl-post-link.**

This seems off. I don't think that those metadata are needed past optimization pipeline in `clang`. With the patch they don't reach `sycl-post-link` at all and the description suggests that they are needed in there.

File: clang/lib/CodeGen/BackendUtil.cpp (hunk 12 above)
Comment (AlexeySachkov) [reply clang/lib/CodeGen/BackendUtil.cpp:1125]: I think that it is also important to note that the pass should be the last in the pipeline. Could be helpful if someone attempts to add a pass which relies on the aforementioned metadata _after_ the cleanup pass

File: llvm/include/llvm/SYCLLowerIR/CleanupSYCLCompilerInternalMetadata.h (hunk 7 above)
Comment (AlexeySachkov) [reply llvm/include/llvm/SYCLLowerIR/CleanupSYCLCompilerInternalMetadata.h:16]: I don't think that we should use `#pragma once` in here: [LLVM Coding Standards](https://llvm.org/docs/CodingStandards.html#c-standard-versions):

> Unless otherwise documented, LLVM subprojects are written using standard C++17 code and avoid unnecessary vendor-specific extensions.

`#pragma once` is **non-standard** C++ functionality, even though widely supported.

I don't see any mentions of it neither in the Coding Standards, nor in [Programmer's Manual](https://llvm.org/docs/ProgrammersManual.html), but there is [Header Guard](https://llvm.org/docs/CodingStandards.html#header-guard) section:

> The header file’s guard should be the all-caps path that a user of this header would `#include`, using ‘_’ instead of path separator and extension marker. For example, the header file `llvm/include/llvm/Analysis/Utils/Local.h` would be `#include`-ed as `#include "llvm/Analysis/Utils/Local.h"`, so its guard is `LLVM_ANALYSIS_UTILS_LOCAL_H`.

Which suggests that macro include guard is the way to go

File: llvm/include/llvm/SYCLLowerIR/CleanupSYCLCompilerInternalMetadata.h [hunk 13]
```diff
@@ -0,0 +1,30 @@
...
+//---------------- CleanupSYCLCompilerInternalMetadata Pass ---------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// Cleanup SYCL compiler internal metadata inserted by the frontend and used by
+// sycl-post-link as it will never be used in the compilation ever again
```
Comment (AlexeySachkov) [inline llvm/include/llvm/SYCLLowerIR/CleanupSYCLCompilerInternalMetadata.h:11]: Once again: removing metadata in `BackendUtils` makes it disappear _before_ `sycl-post-link`

File: llvm/lib/SYCLLowerIR/CleanupSYCLCompilerInternalMetadata.cpp (hunk 9 above)
Comment (AlexeySachkov) [reply llvm/lib/SYCLLowerIR/CleanupSYCLCompilerInternalMetadata.cpp:46]: > I'm not an expert here but usually I see we keep track of if we actually did anything, and if so, we return `PreservedAnalyses::none()`, and if we did nothing, we return `PreservedAnalyses::all()`.

Note: this is a very pessimistic (and probably lazy) approach, which should ideally be avoided. In this case I think it is safe to say that we have preserved all standard LLVM analyses, because we are only removing metadata which we have inserted ourselves, we don't even touch any instructions.

File: llvm/lib/SYCLLowerIR/CleanupSYCLCompilerInternalMetadata.cpp [hunk 14]
```diff
@@ -0,0 +1,47 @@
...
+
+using namespace llvm;
+
+namespace {
+
+void cleanupSYCLCompilerModuleMetadata(const Module &M, llvm::StringRef MD) {
+  NamedMDNode *Node = M.getNamedMetadata(MD);
+  if (!Node)
+    return;
+  Node->clearOperands();
```
Comment (AlexeySachkov) [inline llvm/lib/SYCLLowerIR/CleanupSYCLCompilerInternalMetadata.cpp:28]: That only drops references to those operands, but doesn't remove operands themselves, right?

What happens to such "dead" operands later at `llvm-link` stage?

File: sycl/test/extensions/properties/properties_kernel_device_has.cpp (hunk 1 above)
Comment (AlexeySachkov) [reply sycl/test/extensions/properties/properties_kernel_device_has.cpp:40]: > Well, unfortunately it won't work as opt doesn't provide a way to exclude one pass from the pipeline.

The pass can skip itself if an option is passed. `BackendUtil.cpp` can react to an option to avoid adding a pass into the pipeline at all.

Tagging @steffenlarsen to review the test, he is likely an author of it.

File: sycl/test/optional_kernel_features/atomic_ref-atomic64-aspect.cpp [hunk 15]
```diff
@@ -1,52 +0,0 @@
-// RUN: %clangxx %s -S -o %t.ll -fsycl-device-only
-// RUN: FileCheck %s --input-file %t.ll
-
-// CHECK: !sycl_types_that_use_aspects = !{![[#MDNUM1:]], ![[#MDNUM2:]], ![[#MDNUM3:]]}
```
Comment (AlexeySachkov) [inline sycl/test/optional_kernel_features/atomic_ref-atomic64-aspect.cpp:4]: This is an integration tests which is supposed to check that our headers lead to generation of right LLVM IR. I don't think that just removing them is a good idea. We should use `-disable-llvm-passes` for this (and similar tests) - this should disable the cleanup and the test shouldn't be dependent on any pass, because it is FE who generates that metadata for us

File: llvm/include/llvm/SYCLLowerIR/CleanupSYCLCompilerInternalMetadata.h (hunk 7 above)
Comment (sarnex) [reply llvm/include/llvm/SYCLLowerIR/CleanupSYCLCompilerInternalMetadata.h:16]: Unfortunate, but yeah seems we should use macros, thanks

File: llvm/lib/SYCLLowerIR/CleanupSYCLCompilerInternalMetadata.cpp (hunk 9 above)
Comment (sarnex) [reply llvm/lib/SYCLLowerIR/CleanupSYCLCompilerInternalMetadata.cpp:46]: Thanks, I always saw that but had no idea if it was actually the right thing to do, good to know it's usually not :)

**Comment by bader** [comment]
> > sycl_types_that_use_aspects module-level metadata, so during llvm-link these metadata nodes won't be duplicated significantly increasing module size. **Meanwhile these metadata aren't needed after sycl-post-link.**
> 
> This seems off. I don't think that those metadata are needed past optimization pipeline in `clang`. With the patch they don't reach `sycl-post-link` at all and the description suggests that they are needed in there.

If so, why do we use metadata at all? Can't we use the analysis pass data instead?

**Comment by AlexeySachkov** [comment]
> > > sycl_types_that_use_aspects module-level metadata, so during llvm-link these metadata nodes won't be duplicated significantly increasing module size. **Meanwhile these metadata aren't needed after sycl-post-link.**
> > 
> > 
> > This seems off. I don't think that those metadata are needed past optimization pipeline in `clang`. With the patch they don't reach `sycl-post-link` at all and the description suggests that they are needed in there.
> 
> If so, why do we use metadata at all? Can't we use the analysis pass data instead?

Metadata is emitted to us by FE based on headers content, because we want to control emission of it _without_ compiler modifications, but just by putting a special attribute here and there in header files.

To be more precise here, we have two metadatas which are currently touched by the pass:
- `sycl_aspects` - this metadata communicates to the compiler the exact names of element of `enum aspect` and their assigned values. It solves #5892 for us
- `sycl_types_that_use_aspects` - this metadata communicates a connection between a developer-annotated type and an optional feature it represents. #6970 is an example here

I don't that it is possible to write an analysis pass for the first metadata and for the second it would have to know way too much about the exact spelling/namespaces/mangling of header files to be acceptable

File: sycl/test/extensions/properties/properties_kernel_device_has.cpp (hunk 1 above)
Comment (MrSidims) [reply sycl/test/extensions/properties/properties_kernel_device_has.cpp:40]: I ended up adding -Xclang -disable-llvm-passes to the tests as Alexey has suggested below

File: clang/lib/CodeGen/BackendUtil.cpp (hunk 4 above)
Comment (MrSidims) [reply clang/lib/CodeGen/BackendUtil.cpp:46]: Applied, thanks!

File: clang/lib/CodeGen/BackendUtil.cpp (hunk 12 above)
Comment (MrSidims) [reply clang/lib/CodeGen/BackendUtil.cpp:1125]: > I think being specific is helpful in the comments. It might also make sense to elaborate on why we are removing them

TBH, like now the pass removes only module level metadata, but later I can image it will start removing more

> I think that it is also important to note that the pass should be the last in the pipeline.

Indeed, applied

File: llvm/lib/SYCLLowerIR/CleanupSYCLCompilerInternalMetadata.cpp (hunk 14 above)
Comment (MrSidims) [reply llvm/lib/SYCLLowerIR/CleanupSYCLCompilerInternalMetadata.cpp]: That's a funny question, in IR I see no dead nodes and everything works just fine, while the description and implementation of clearOperands indeed just drops references. I'll double check tomorrow.

File: sycl/test/optional_kernel_features/atomic_ref-atomic64-aspect.cpp (hunk 15 above)
Comment (MrSidims) [reply sycl/test/optional_kernel_features/atomic_ref-atomic64-aspect.cpp:4]: Added -disable-llvm-passes, thanks!

File: llvm/lib/SYCLLowerIR/CleanupSYCLCompilerInternalMetadata.cpp (hunk 14 above)
Comment (MrSidims) [reply llvm/lib/SYCLLowerIR/CleanupSYCLCompilerInternalMetadata.cpp]: If I read the code correctly we will end up calling d'tor for md operands:
[SmallVectorTemplateBase](https://llvm.org/doxygen/classllvm_1_1SmallVectorTemplateBase.html#a59bf49ac292d1f2d6ab5db4be8d98f3d)(size_t [Size](https://llvm.org/doxygen/classllvm_1_1SmallVectorBase.html#a7b5a03b19133c790a4d6fff66a5d2135)) : [SmallVectorTemplateCommon](https://llvm.org/doxygen/classllvm_1_1SmallVectorTemplateCommon.html)<[T](https://llvm.org/doxygen/classT.html)>([Size](https://llvm.org/doxygen/classllvm_1_1SmallVectorBase.html#a7b5a03b19133c790a4d6fff66a5d2135)) {}
 
  static void [destroy_range](https://llvm.org/doxygen/classllvm_1_1SmallVectorTemplateBase.html#a2d25f6f9379888156d4dc87f0c70a0da)([T](https://llvm.org/doxygen/classT.html) *S, [T](https://llvm.org/doxygen/classT.html) *[E](https://llvm.org/doxygen/BuiltinGCs_8cpp.html#ab5899ca200d34e3cc6bb09ebce5e5b3c)) {
    while (S != [E](https://llvm.org/doxygen/BuiltinGCs_8cpp.html#ab5899ca200d34e3cc6bb09ebce5e5b3c)) {
      --[E](https://llvm.org/doxygen/BuiltinGCs_8cpp.html#ab5899ca200d34e3cc6bb09ebce5e5b3c);
      [E](https://llvm.org/doxygen/BuiltinGCs_8cpp.html#ab5899ca200d34e3cc6bb09ebce5e5b3c)->~T();
    }
  }

so the operands are not "dead", but dead.

File: clang/lib/CodeGen/BackendUtil.cpp [hunk 16]
```diff
@@ -1124,6 +1125,10 @@ void EmitAssemblyHelper::RunOptimizationPipeline(
       if (LangOpts.SYCLIsNativeCPU) {
         MPM.addPass(PrepareSYCLNativeCPUPass());
       }
+
+      // Remove SYCL metadata added by the frontend, like sycl_aspects
+      // Note, this pass should be in the end of the pipeline
```
Comment (sarnex) [inline clang/lib/CodeGen/BackendUtil.cpp:1130]: nit

```suggestion
      // Note, this pass should be at the end of the pipeline
```

File: llvm/lib/SYCLLowerIR/CleanupSYCLMetadata.cpp [hunk 17]
```diff
@@ -0,0 +1,45 @@
...
+  Node->dropAllReferences();
+  Node->eraseFromParent();
+}
+
+} // anonymous namespace
+
+PreservedAnalyses CleanupSYCLMetadataPass::run(Module &M,
+                                               ModuleAnalysisManager &MAM) {
+  // Remove SYCL module-level metadata that will never be used again to avoid
+  // its duplication of their operands during llvm-link hence preventing
```
Comment (sarnex) [inline llvm/lib/SYCLLowerIR/CleanupSYCLMetadata.cpp:37]: nit

```suggestion
  // duplication of their operands during llvm-link hence preventing
```

**Review by sarnex — APPROVED** [review]
LGTM, thanks!

File: llvm/lib/SYCLLowerIR/CleanupSYCLCompilerInternalMetadata.cpp (hunk 14 above)
Comment (AlexeySachkov) [reply llvm/lib/SYCLLowerIR/CleanupSYCLCompilerInternalMetadata.cpp]: Thanks!

I would suggest to add a test (could be an update to any existing one) which makes it so that one of operands of removed metadata is also an operand of a non-removed metadata. To make sure that it doesn't crash, at least. Because with simple metadata as `!N = !{i32 4}`, there is very high chance that it is used by more than one metadata node in a bigger application.

**Review by AlexeySachkov — APPROVED** [review]
LGTM

**Review by bso-intel — APPROVED** [review]

**Comment by MrSidims** [comment]
@intel/llvm-gatekeepers please help with merge. clang-format issues are related with community code
